# QUESTION 1 Final exam

In [1]:
import numpy as np
from sklearn.svm import SVC

Create a helper function to load data from csv file (cancer_data.csv)

In [15]:
def loadDataSetFromFile(fileName):
    dataMat = []; labelMat = []
    fr = open(fileName)
    #fr.readlines() #skip header
    for line in fr.readlines()[1:]:  #skip the hearder
        lineArr = line.strip().split(',') #strip the new lines and split the line by using ',' as delimer
        #print(lineArr)
        #radius= lineArr[0]) texture=lineArr[1] etc... fractal dimension =  lineArr[9]  and cancer=ineArr[-1] last line
        dataMat.append([float(lineArr[0]), float(lineArr[1]) , float(lineArr[2]) , float(lineArr[3]), float(lineArr[4]), float(lineArr[5]), float(lineArr[6]), float(lineArr[7]), float(lineArr[8]), float(lineArr[9])])
        labelMat.append(float(lineArr[-1]))  # take the last column (cancer) : the target variable
    return dataMat,labelMat

Load the data in memory from cancer_data.csv file

In [16]:
dataMat,labelMat = loadDataSetFromFile('cancer_data.csv')

Transform the list of lists into 2D array for the algorithm

In [17]:
dataMatArray = np.array(dataMat)

Transform the list into array

In [18]:
labelMatArray= np.array(labelMat)

Create a Support Vector Classification model object

In [19]:
clf1 = SVC()

Fit the model

In [20]:
clf1.fit(dataMatArray, labelMatArray) 

SVC(C=1.0, cache_size=200, class_weight=None, coef0=0.0,
  decision_function_shape=None, degree=3, gamma='auto', kernel='rbf',
  max_iter=-1, probability=False, random_state=None, shrinking=True,
  tol=0.001, verbose=False)

Predict a new patient

## Predict an observation

   - radius: 14
   - texture: 14
   - perimeter: 88
   - area: 566
   - smoothness: 1
   - compactness: 0.08
   - concavity: 0.06
   - concae points: 0.04
   - symmetry: 0.18
   - fractal dimension: 0.05

In [21]:
  pred_patient = clf1.predict([[14, 14, 88, 566, 1, 0.08, 0.06, 0.04,0.18, 0.05]])

In [22]:
  pred_patient 

array([ 1.])

#### Response: The patient is in class 1

# QUESTION 2 FINAL

In [23]:
from sklearn.datasets import fetch_20newsgroups
from sklearn.feature_extraction.text import CountVectorizer, TfidfTransformer
from sklearn.pipeline import Pipeline
#from sklearn_linear_model import SGDClassifier   error
from sklearn.linear_model import SGDClassifier

categories = [ 'rec.sport.baseball','rec.sport.hockey']
#twenty_train = fetch_20newsgroups(subset='test', categories=categories) error
twenty_train = fetch_20newsgroups(subset='train', categories=categories)
#twenty_test = fetch_20newsgroups(subset='train', categories=categories, shuffle=True, random_state=42) error
twenty_test = fetch_20newsgroups(subset='test', categories=categories, shuffle=True, random_state=42)

#error Pipeline instead of pipeline Capitalcase for p
text_clf = Pipeline([('vect', CountVectorizer()),
                     ('tfidf', TfidfTransformer()),
                     ('clf', SGDClassifier(loss='hinge', penalty='l2',
                                           alpha=1e-3, random_state=42)),
])
#text_clf.fit(twenty_train.data, twenty_test.target)  error
text_clf.fit(twenty_train.data, twenty_train.target)  
predicted = text_clf.predict(twenty_test.data)

## Errors

- Pipeline instead of pipeline
- change test to train in twenty_train creattion
- Change train to test in twnety_test creation
- in model fitting use twenty_train.target instead of twenty_test.target
- Import SGDClassifier from sklearn.linear_model instead of sklearn.linear_model

 ### 1. How many observations are in the training dataset?

In [24]:
count_vect =  CountVectorizer ()
X_train_counts = count_vect.fit_transform (twenty_train.data)
X_train_counts.shape

(1197, 18571)

#### There are 1197 observations in the trainning dataset

### 2. How many features are in the training dataset?

#### There are 18571 features in the trainning dataset

### 3. How well did your model perform?

In [25]:
from sklearn.metrics import mean_squared_error, r2_score

In [26]:
mse = mean_squared_error(twenty_test.target, predicted)

In [27]:
mse

0.031407035175879394

In [28]:
r2_score(twenty_test.target, predicted)

0.87437106620455418

#### R2 =  87.43% means 87.43 of the variabilty of the target variable is explained by the model. It is very good but it depend of the  level of precision we are looking.

In [30]:
from sklearn.metrics import classification_report
print(classification_report(predicted, twenty_test.target))

             precision    recall  f1-score   support

          0       0.96      0.97      0.97       394
          1       0.97      0.97      0.97       402

avg / total       0.97      0.97      0.97       796



#### The recall (around 97%) are vey good for the 2 classes and the presision (around 97%) as well. That is means false ngative and false positive are very low.

#### Which category is Hockey? 0 or 1

In [29]:
pred_hokey = text_clf.predict(["Hockey"])

In [30]:
pred_hokey

array([1])

#### Response: Hokey is in category  0

#### Which category is baseball? 0 or 1

In [31]:
pred_baseball = text_clf.predict(["baseball"])

In [32]:
pred_baseball

array([0])

#### Response: baseball is in in category  0